### ISSUE TITLE
Bind variable in `With` block to a name

### ISSUE THREAD

**Issue by Greedquest** [twinbasic/lang-design#57]
## Motivation
I use `With` blocks like _context managers_ to ensure a temporary object gets cleaned up when I'm done with it. With a regular `With` block I often find I would like to refer to the variable itself particularly when passing it to some sub-function.  E.g.

```vba
With New Scripting.FileSystemObject
	Dim files As Variant = .Find("*.xlam")
	doMoreProcessing(files, /* pass fso here */)
End With
```

You often see this pattern with `.Self`
```vba
With New FileWriter("C.txt")
	'do stuff
	writeHeader .Self
	writeBody .Self
End With 'closes file
```
Of course that only works if the object is under your control.

If I declare the variable eagerly
```vba
Dim fso As FileSystemObject = New FilesystemObject
With fso
	Dim files As Variant = .Find("*.xlam")
	doMoreProcessing(files, fso)
End With
```
Then we don't actually clear up the object on `End With` which defeats the purpose. 

## Suggested Syntax
```vba
With New Excel.Application Alias app
	app.Name
End With

With CreateObject("Excel.Application") Alias app
	app.Name
End With
```
_Note I use `CreateObject` in these examples but really any expression that returns an Object would be allowed_

### Semantics
The scope of the name `app` will be only within the `With` block. Also once `Alias`'ed the variable cannot be used anonymously with just `.foo`

```vba
With New Excel.Application Alias app
	app.Name 'fine
	.Name 'now an error with the alias specified the variable name must be used
End With
app.Name 'not fine, app was defined in the block only
``` 
Nested with blocks get access to all their parent variables

```vba
With Sheet1.Range("A1") Alias cellToFill
	With cellToFill.Interior Alias cellInterior
		cellInterior.ColorIndex = Abs(cellToFill.Value)
		cellInterior.InvertIfNegative = True
	End With
	processMore cellToFill
End With
```

### Alternatives Considered

```vba
With app As New Excel.Application 'confusion with auto instantiated variable
	app.Name
End With

With app As CreateObject("Excel.Application") 'confusion as we expect to see a type here, not an expression
	app.Name
End With
```

```vba
With New Excel.Application As app 'reads nicely but inverted compared to what we usually see
	app.Name
End With

With CreateObject("Excel.Application") As app 'looks quite nice but inverted as above
	app.Name
End With
```

```vba
With app As Excel.Application = New Excel.Application 'quite verbose
	'plus, in standard With blocks the type of the expression is always known so shouldn't be specified manually
	app.Name
End With

With app As Object = CreateObject("Excel.Application") 'again the `As Object` gives no information
	app.Name
End With

'Also those 2 forms imply the existence of a late bound implicit variant form:
With app = CreateObject("Excel.Application") 'this form is ambiguous, it could evaluate to boolean
	app.Name
End With
```

**Comment by wqweto** [comment]
Keep in mind that the implicit variable has a reference type the compiler is aware of. It's not a `Variant` and it's not simply an `Object` as it inherits the data-type of the expression used in `With` initialization.

I would not introduce an explicit alias for the hidden variable but allow addressing it with some new/weird syntax something like `.0` for current `With` block variable and `.1` for parent one etc. and keep "bare" syntax of `.Prop/Method` executing fast i.e. try to keep the implicit variable in a register for fast(er) member access.

**Comment by mburns08109** [comment]
@wqweto,

.0, .1, etc.? Oh I hate that idea - it smacks too much of "secret code magic" stuff!

How about something more intuitive and within a WITH block like ".With" maybe if we really need to go this direction (which, I can admit, will allow some easier custom cleanup of COM Refcount issues in some circumstances - especially for imported VB6 legacy code)?

If there is a need to access the implicit With variable for surrounding/nested With...End With blocks, well you're just going to need to declare them and use actual variables for that. Don't be so danged lazy about this! Remember KISS principles? We hold to those for a good reason in most circumstances. Why is it that the first answer always seems to be "make the magic happen for me"? Sorry, but coding is work. Declare a damned variable if you need to use it in a nested With---End With structure.

This is BASIC, not a C++/C# variant language!

**Comment by mwolfe02** [comment]
A couple of other alternatives:

```vba
With New Excel.Application Dim app
    app.Name
End With

With CreateObject("Excel.Application") Dim app
    app.Name
End With
```

`Alias` is not bad, but I think the `Dim` keyword is more semantically similar to what we're actually doing here.

If twinBASIC adds [support for block-level scoping](https://github.com/twinbasic/twinbasic/discussions/1463), I think the `Local` keyword works even better:

```vba
With New Excel.Application Local app
    app.Name
End With

With CreateObject("Excel.Application") Local app
    app.Name
End With
```

**Comment by mansellan** [comment]
I would be strongly against any syntax that puts the type ahead of the variable. Nothing else in VBx does this:

```vb
With New MyClass Local x
```

It's unconventional, and requires a different mode of thinking. It's cognitive friction.

```vb
With app As MyClass = New MyClass()
```

(parens optional) may be verbose, but it fits with the language, including the new As New vs = New syntax.

It can cope with constructors:

```vb
With app As MyClass = New MyClass(1, "Products")
```
 Further, in v.Next, with type inference, it could evolve to:

```vb
With app As MyClass = New(1, "Products")
```

**Comment by mojear** [comment]
In the traversal loop in Autohotkey, they do this:

```
Loop 3
{
      ; A_Index will be 1, 2, then 3
      MsgBox "Iteration number is " A_Index 
      Sleep 100
}

```
"A_Index" is a special variable that they reserve for this environment, indicating the index being traversed.

Why can't we use a special variable to indicate the object being With?

**Comment by WaynePhillipsEA** [comment]
Oh we can do that.   But apparently we shouldn't be introducing any new syntax at this stage (...see your own comment here https://github.com/twinbasic/twinbasic/issues/1972#issuecomment-2562335860).  🤷‍♂️🤷‍♂️

Any such change has consequences that need to be considered.  In this case, ensuring that a chosen symbol name cannot conflict with any other part of any existing VBx code in existence.  The easiest way to do this is to make the new symbol be exposed in a way that is considered invalid-syntax in traditional VBx code.  That's why this feature hasn't been implemented yet - we need a proper proposal for a suitable syntax, and one that fits nicely alongside existing BASIC syntax.

**Comment by mojear** [comment]
> Oh we can do that. But apparently we shouldn't be introducing any new syntax at this stage (...see your own comment here [twinbasic/twinbasic#1972 (comment)](https://github.com/twinbasic/twinbasic/issues/1972#issuecomment-2562335860)). 🤷‍♂️🤷‍♂️
> 
> Any such change has consequences that need to be considered. In this case, ensuring that a chosen symbol name cannot conflict with any other part of any existing VBx code in existence. The easiest way to do this is to make the new symbol be exposed in a way that is considered invalid-syntax in traditional VBx code. That's why this feature hasn't been implemented yet - we need a proper proposal for a suitable syntax, and one that fits nicely alongside existing BASIC syntax.

My words may mislead you.

I am also currently doing a text work related to the new and old versions.

When I am working on the old version, I will treat the new version as if it does not exist. My goal is to make it 100% compatible with the old version. Clear goals and clear purposes. I will move towards that goal.

Otherwise, I will fall into a quagmire.

As a person who has benefited from this practice, I recommend that TwinBasic open a new and detailed section so that those who are looking forward to the new syntax can speak freely in it.

Just use it as a message board, a wish list for new syntax and new features to collect user needs. TwinBasic does not need to interact in that section, and does not need to care whether the suggestions put forward by users are feasible. TwinBasic will treat that section as if it does not exist and does not need to browse.

When you are done with the current goal, go there to check the suggestions left by users.

Adding new syntax should be done carefully and carefully, which is related to the future of this language. It not only includes compatibility, but also whether the style is consistent, the advantages and disadvantages of each party, etc. The addition of += and -= may determine the life and death of this language. Users can leave some opinions in the forum, and when the TwinBasic team makes decisions, those opinions will also serve as data support for decision-making.

At the same time, including documents, interactions, etc., should have sections, which are also considered part of the language, and users' opinions should also be listened to and saved there.

In short, I think TwinBasic should have a new section to separate the current priority tasks from the tasks that may be used in the future but are the least important at the moment.

If I were you, I would not add any new syntax to V1, not even AndAlso, just 100% compatible with VBx, just this one task. First, let old users migrate their projects to TwinBasic as soon as possible. And promise that VNext will add new syntax.

Put the new syntax as a question, first let everyone answer, think, and argue. Finally, when you have free time, you can weigh the pros and cons and choose whether to add it.

**Comment by WaynePhillipsEA** [comment]
I suggest you take a look at the New Features documentation:

https://github.com/twinbasic/documentation/wiki/twinBASIC-Features

This project is continuously pushing forwards.   I'm afraid you'll have to accept it, or move on to another language.

**Comment by FullValueRider** [comment]
Again Ooop steps into the breach.   There is no need to add syntax to the language when you can create a class that has knowledge of how other entities are accessed (whether its a dictionary, collection, arraylist, stack,queue , long string, etc).  In mycase the Class is called EnumIt.

The Enumit class has a number of methods
• HasNext
• HasPrev
• MoveNext - returns true if the move to the next item was successful
• MovePrev - returns true if the move to the previous item was successful
• MoveToStart - Move to the position defined as start by ay FTS
• MoveToEnd - Move to the end position defined by any FTS
• MoveToFirst - Move to the first index
• MoveToLast - Move to the last index
• FTS( fromindex, toIndex, Step, Rank) - for multidimension arrays FTS sets the start/end/step for all ranks or individual ranks (and negative indices mean count from RHS rather than LSH)
• It (short for of Item(0) - the current item
• Item(offset index) - Read/Write property
• Key(offset index)  - Key is defined wider than the meaning of key in a dictionary Read property
• Offset() - from first index
EnumIt is agnostic to the type it receives (if you give it a long it will enumerate once, if you give it a string it will enumerate each character int the string)

so now I can write
```VB
    Dim myE as EnumIt = EnumIt.Deb(ThingToEnumerate)  ' Deb is the self factory method for the predecalredId.
    Do
            Debug.print  myE.It          ' the current item - short for .Item(0)
            Debug.Print  myE.Item(3)  ' the Item at position 3 from the current item
            Debug.Print myE.Key        ' The key or Index of the current item 
            Debug.Print myE.Offset    ' the offset of the current item from the first index
     Loop while myE.MoveNext     '
     
  ```
And because enumIt is a class, you can enumerate multiple variables at the same time, using the same syntax.

**Comment by mburns08109** [comment]
@Greedquest 

re:  > Also once Alias'ed the variable cannot be used anonymously with just .foo

Would that not defeat the very purpose of a WITH block? What's your reasoning for this?

**Comment by Greedquest** [comment]
> @Greedquest 
> 
> re:  > Also once Alias'ed the variable cannot be used anonymously with just .foo
> 
> Would that not defeat the very purpose of a WITH block? What's your reasoning for this?


The With block can be useful to make shorthand expressions but honestly my main use is generally to make temporary objects and ensure cleanup class_terminate code is called as early as possible (see the 2 examples in my initial post).

The tricky question is the scope of these variables, in my opinion they would be block scoped so can't be called from outside the with block they are declared in, similar to the anonymous version. But we might want to mandate a Local keyword for that.


For tB v2+ we could also introduce Class_Enter and Class_Exit events that are called in case we aren't using the with block for a temporary variable but for something whose terminate event won't necessarily be called after the with block. Ideally these events would occur even on error. We also could even have `With a As Long, b as Integer` for arbitrary block scoped variables.

This would make resource cleanup more predictable.
